apps/OboeTester/app/src/main/java/com/mobileer/oboetester/TestAudioActivity.java

```
@@ -820,43 +941,93 @@ protected void toastPauseError(int result) {
         } else {
-            mAudioState = AUDIO_STATE_PAUSED;
             updateEnabledWidgets();
             onStopAllContexts();
         }
     }
 
     public void flushAudio() {
-        int result = flushNative();
+        int result;
+        synchronized (mAudioStateLock) {
+            if (mAudioState != AUDIO_STATE_PAUSED) {
+                Log.w(TAG, "flushAudio() called in invalid state " + mAudioState + ", ignoring.");
+                return;
+            }
+            result = flushNative();
+            if (result == 0) {
+                mAudioState = AUDIO_STATE_FLUSHED;
+            }
+        }
+
         if (result != 0) {
             showErrorToast("Flush failed with " + result + ", " + convertErrorToText(result));
         } else {
-            mAudioState = AUDIO_STATE_FLUSHED;
             updateEnabledWidgets();
         }
     }
 
     public void stopAudio() {
-        int result = stopNative();
+        synchronized (mAudioStateLock) {
+            stopAudioLocked();
+        }
+    }
+
+    @GuardedBy("mAudioStateLock")
+    private void stopAudioLocked() {
+        int result;
+        if (mAudioState != AUDIO_STATE_STARTED && mAudioState != AUDIO_STATE_PAUSED) {
+            Log.w(TAG, "stopAudio() called in invalid state " + mAudioState + ", ignoring.");
+            return;
+        }
+        result = stopNative();
+        if (result == 0) {
+            mAudioState = AUDIO_STATE_STOPPED;
+        }
+
         if (result != 0) {
             showErrorToast("Stop failed with " + result + ", " + convertErrorToText(result));
         } else {
-            mAudioState = AUDIO_STATE_STOPPED;
             updateEnabledWidgets();
             onStopAllContexts();
         }
     }
 
     public void releaseAudio() {
-        int result = releaseNative();
+        int result;
+        synchronized (mAudioStateLock) {
+            if (mAudioState < AUDIO_STATE_OPEN || mAudioState >= AUDIO_STATE_RELEASED) {
```

Don't check the state and silence the failure here. Reason as above